src/setuptools_scm/config.py

```diff
@@ -0,0 +1,92 @@
+""" configuration """
+from __future__ import print_function, unicode_literals
+import os
+import re
+import warnings
+
+from .utils import trace
+
+DEFAULT_TAG_REGEX = r'^(?:\w+-)?(?P<version>v?\d+(?:\.\d+){0,2}[^\+]+)(?:\+.*)?$'
+DEFAULT_VERSION_SCHEME = 'version_scheme'
+
+
+def _check_tag_regex(value):
+    if not value:
+        value = DEFAULT_TAG_REGEX
+    regex = re.compile(value)
+
+    group_names = regex.groupindex.keys()
+    if regex.groups == 0 or (regex.groups > 1 and 'version' not in group_names):
+        warnings.warn("Expected tag_regex to contain a single match group or a group named 'version' " +
+                      "to identify the version part of any tag.")
+
+    return regex
+
+
+def _check_absolute_root(root, relative_to):
+    if relative_to:
+        if os.path.isabs(root) and not root.startswith(relative_to):
+            warnings.warn("absolute root path '%s' overrides relative_to '%s'" % (root, relative_to))
+        root = os.path.join(os.path.dirname(relative_to), root)
+    return os.path.abspath(root)
+
+
+class Configuration(object):
```

Comment:
part of me wonders if it shouldn't be ok to just make Configuration immutable
in order to avoid having a set of properties that need to be smart,

this thought is for a followup however and mainly food forthought

Reply:
right now i'm not happy with this config class, but its working and adds the feature needed so i'm happy to merge, its up to me or someone else to make me more happy with it afterwards